symbolic-debuginfo/src/pdb.rs

```diff
@@ -409,119 +419,102 @@ impl<'data, 'object> Iterator for PdbSymbolIterator<'data, 'object> {
-
 struct PdbStreams<'d> {
     debug_info: Arc<pdb::DebugInformation<'d>>,
     type_info: pdb::TypeInformation<'d>,
     id_info: pdb::IdInformation<'d>,
+    string_table: Option<pdb::StringTable<'d>>,
+
+    pdb: Arc<RwLock<Pdb<'d>>>,
+
+    /// ModuleInfo objects are stored on this object (outside PdbDebugInfo) so that the
+    /// PdbDebugInfo can store a TypeFormatter, which has a lifetime dependency on its
+    /// ModuleProvider, which is this PdbStreams. This is so that TypeFormatter can cache
+    /// CrossModuleImports inside itself, and those have a lifetime dependency on the
+    /// ModuleInfo.
+    module_infos: FrozenMap<u16, Box<ModuleInfo<'d>>>,
 }
 
 impl<'d> PdbStreams<'d> {
     fn from_pdb(pdb: &PdbObject<'d>) -> Result<Self, PdbError> {
         let mut p = pdb.pdb.write();
 
+        // PDB::string_table errors if the named stream for the string table is not present.
+        // However, this occurs in certain PDBs and does not automatically indicate an error.
+        let string_table = match p.string_table() {
+            Ok(string_table) => Some(string_table),
+            Err(pdb::Error::StreamNameNotFound) => None,
+            Err(e) => return Err(e.into()),
+        };
+
         Ok(Self {
+            string_table,
             debug_info: pdb.debug_info.clone(),
             type_info: p.type_information()?,
             id_info: p.id_information()?,
+            pdb: pdb.pdb.clone(),
+            module_infos: FrozenMap::new(),
         })
     }
+}
 
-    fn type_map(&self) -> TypeMap<'_> {
-        ItemMap {
-            iter: self.type_info.iter(),
-            finder: self.type_info.finder(),
+impl<'d> pdb_addr2line::ModuleProvider<'d> for PdbStreams<'d> {
+    fn get_module_info(
+        &self,
+        module_index: u16,
+        module: &Module,
+    ) -> Result<Option<&ModuleInfo<'d>>, pdb::Error> {
+        if let Some(module_info) = self.module_infos.get(&module_index) {
+            return Ok(Some(module_info));
         }
-    }
 
-    fn id_map(&self) -> IdMap<'_> {
-        ItemMap {
-            iter: self.id_info.iter(),
-            finder: self.id_info.finder(),
-        }
+        let mut pdb = self.pdb.write();
+        Ok(pdb.module_info(module)?.map(|module_info| {
+            self.module_infos
+                .insert(module_index, Box::new(module_info))
+        }))
     }
 }
 
 struct PdbDebugInfo<'d> {
-    /// The original PDB to load module streams on demand.
-    pdb: Arc<RwLock<Pdb<'d>>>,
+    /// The streams, to load module streams on demand.
+    streams: &'d PdbStreams<'d>,
     /// All module headers for repeated iteration.
-    modules: Vec<Module<'d>>,
-    /// Lazy loaded module streams in the same order as headers.
-    module_infos: Vec<LazyCell<Option<ModuleInfo<'d>>>>,
-    /// Cache for module by name lookup for cross module imports.
-    module_exports: RefCell<BTreeMap<pdb::ModuleRef, Option<pdb::CrossModuleExports>>>,
+    modules: Rc<Vec<Module<'d>>>,
```

Comment:
Let me double-check if its okay to make this `!Send` (even though the `RefCell`s previously made this `!Sync`).

(This also makes me wonder why `TypeFormatter` needs a `Rc<Vec<_>>` and is not satisfied with a `&[_]` instead.)

Reply:
I took another look at this. A slice is indeed not an option: Let's say it was `&'list [Module<'dbi>]`, what would those two lifetimes be? Who would own the `Vec`, and who would own the `DebugInformation`?

But at least I've eliminated the `Rc` now:

I've changed `Rc<Vec<Module<'d>>` to `Vec<Module<'d>>`, i.e. I've given the `TypeFormatter` full ownership of the `Vec` instead of just shared ownership.
Now everyone who wants to access the list of modules needs to go through the `TypeFormatter` - `TypeFormatter::modules()` exposes the slice.